File: adventofcode/2022/day15/day15.rb
# frozen_string_literal: true

def read_input
  ARGF.readlines
end

def mdist p1, p2
  x1, y1 = p1
  x2, y2 = p2
  (x1 - x2).abs + (y1 - y2).abs
end

def parse_input lines
  lines.reduce({}) do |s, line|
    if /Sensor at x=(-?\d+), y=(-?\d+): closest beacon is at x=(-?\d+), y=(-?\d+)/ =~ line
      sp = [$1.to_i, $2.to_i]
      bp = [$3.to_i, $4.to_i]
      s[sp] = {
        bpos: bp,
        range: mdist(sp, bp),
      }
      s
    end
  end
end

def cannot_have_beacon? sensors, point
  return false if sensors.has_key?(point)
  return false if sensors.any? {|_, v| v[:bpos] == point}

  sensors.any? do |p, b|
    mdist(p, point) <= b[:range]
  end
end

def count_pos_without_beacon sensors, xmm, y
  xmin, xmax = xmm
  (xmin..xmax).count { |xi| cannot_have_beacon?(sensors, [xi, y]) }
end

def part1 sensors
  xmm = sensors.flat_map {|p, b| [p[0], p[0] - b[:range], p[0] + b[:range]]}.minmax
  count_pos_without_beacon sensors, xmm, 2000000
end

def tuning_frequency point
  x, y = point
  x * 4000000 + y
end

def part2 sensors
  sensors.each do |sp, b|
    sx, sy = sp
    # search for points range+1 distance away from sensors
    (0..(b[:range]+1)).each do |dx|
      dy = b[:range] + 1 - dx

      [[-1, -1], [-1, 1], [1, -1], [1, 1]].each do |signx, signy|
        x, y = [sx + (dx * signx), sy + (dy * signy)]
        point = [x, y]

        # next unless (0..20).include?(x) && (0..20).include?(y)
        next unless (0..4000000).include?(x) && (0..4000000).include?(y)

        return tuning_frequency(point) if sensors.all? { |p, b| mdist(p, point) > b[:range] }
      end
    end
  end
end

sensors = parse_input read_input
p part1 sensors
p part2 sensors

File: adventofcode/2022/day15/day15.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
)

var sensorRegex = regexp.MustCompile(`Sensor at x=(-?\d+), y=(-?\d+): closest beacon is at x=(-?\d+), y=(-?\d+)`)

func readInput() []string {
	lines := []string{}
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines
}

func parseSensor(line string) Sensor {
	caps := sensorRegex.FindStringSubmatch(line)
	sx, _ := strconv.Atoi(caps[1])
	sy, _ := strconv.Atoi(caps[2])
	bx, _ := strconv.Atoi(caps[3])
	by, _ := strconv.Atoi(caps[4])
	spos := Point{x: sx, y: sy}
	bpos := Point{x: bx, y: by}
	return Sensor{
		pos:   spos,
		bpos:  bpos,
		mdist: mdist(spos, bpos),
	}
}

func parseInput(lines []string) map[Point]Sensor {
	res := make(map[Point]Sensor)
	for _, line := range lines {
		s := parseSensor(line)
		res[s.pos] = s
	}
	return res
}

type Point struct {
	x, y int
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

func mdist(p1 Point, p2 Point) int {
	return abs(p1.x-p2.x) + abs(p1.y-p2.y)
}

type Sensor struct {
	pos   Point
	bpos  Point
	mdist int
}

func no_beacon(sensors map[Point]Sensor, point Point) bool {
	if _, ok := sensors[point]; ok {
		return false
	}

	for p, s := range sensors {
		if point == s.bpos {
			return false
		}

		if mdist(p, point) <= s.mdist {
			return true
		}
	}
	return false
}

func count_point_no_beacon(sensors map[Point]Sensor, xmin, xmax, y int) int {
	cnt := 0
	for i := xmin; i <= xmax; i++ {
		p := Point{x: i, y: y}
		if no_beacon(sensors, p) {
			cnt += 1
		}
	}
	return cnt
}

func part1(sensors map[Point]Sensor) int {
	xmin := 0
	xmax := 0
	for p, s := range sensors {
		if xmin > p.x-s.mdist {
			xmin = p.x - s.mdist
		}
		if xmax < p.x+s.mdist {
			xmax = p.x + s.mdist
		}
	}
	return count_point_no_beacon(sensors, xmin, xmax, 2000000)
}

func tuning_frequency(p Point) int64 {
	return int64(p.x)*4000000 + int64(p.y)
}

func part2(sensors map[Point]Sensor) int64 {
	pchan := make(chan Point)
	pdiffs := [][]int{{-1, -1}, {-1, 1}, {1, -1}, {1, 1}}
	for _, s := range sensors {
		go func(c chan Point, s Sensor) {
			for dx := 0; dx <= (s.mdist + 1); dx++ {
				dy := s.mdist + 1 - dx
				for _, sxy := range pdiffs {
					x := s.pos.x + dx*sxy[0]
					y := s.pos.y + dy*sxy[1]
					point := Point{x, y}

					if x < 0 || x > 4000000 || y < 0 || y > 4000000 {
						continue
					}

					if !no_beacon(sensors, point) {
						c <- point
					}
				}
			}
		}(pchan, s)
	}
	point := <-pchan
	return tuning_frequency(point)
}

func main() {
	sensors := parseInput(readInput())
	fmt.Println(part1(sensors))
	fmt.Println(part2(sensors))
}
